keps/sig-etcd/4578-server-feature-gate/README.md

```diff
@@ -0,0 +1,331 @@
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[etcd-io/website]: https://github.com/etcd-io/website
+
+## Summary
+
+We are introducing the Kubernetes style feature gate framework into etcd to gate future feature enhancement behind a sequence of feature lifecyles. Users would be able to turn features on or off and query feature enablement of an etcd server in a consistent way. 
+
+## Motivation
+
+Currently any new enhancements to the etcd are typically added as [an experimental feature](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/features.md#adding-a-new-feature), with a configuration flag prefixed with "experimental", e.g. --experimental-feature-name. 
+
+When it is time to [graduate an experimental feature to stable](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/features.md#graduating-an-experimental-feature-to-stable), a new stable feature flag identical to the experimental feature flag but without the --experimental prefix is added to replace the old feature flag, which is a breaking change.
+
+The following problems exist with the current `--experimental` flag mechanism: 
+* hard to track when and why feature was introduced, and what stage it is in.
+* no clear path to graduate flags, as removal of experimental prefix is a breaking change in command line.
+* unclear, undocumented feature flag graduation criteria. So very often features get stuck or forgotten in experimental phase.
+* no method for clients to query and make decisions based on feature enablement in a cluster.
+* each `--experimental` flag will have to be piped through the code based separately. It is cumbersome, nothing stops them being modified somewhere along the path, and some code changes might slip through and not be guarded by the flag.
+
+We are proposing to introduce Kubernetes style [feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) into etcd to address these issues.
+
+### Goals
+
+* Introduce a single mechanism in the code base for enabling/disabling features with code changes guarded behind the mechanism without risking breaking the rest of the server if the feature is disabled.
+
+* Introduce a way for a client to find out what features are enabled in the server.
+
+* Codify lifecycle stages of a feature, and clear tracking and documentation of feature lifecycle progression.
+
+### Non-Goals
+
+* Handle the feature gate enablement on the cluster level. It would be designed in a separate KEP.
+
+* Migrate all existing `--experimental` features to feature gate. 
+  * While feature gate provides a [path to deprecate existing `--experimental` features]((#path-to-migrate-existing-experimental-features)), and we will migrate several `--experimental` during the feature gate implementation as examples, the completion of the KEP does not require migrating all existing `--experimental` features. It can be done as followup tasks.
```

Comment:
Don't exactly agree, we shouldn't add a second standard for managing features without having a plan for the first one. This KEP should propose a plan to migrate (and not leave any flags behind) to feature gates and execution of this should be done as part of the KEP effort.

Reply:
ok. Added migration as part of the goal, and Milestone 3.